### ISSUE TITLE
Issues with backtracking line search

### ISSUE THREAD

**Issue by w1th0utnam3** [argmin-rs/argmin#154]
I noticed some points while playing around with LBFGS in combination with backtracking line search and would like to hear your thoughts before submitting a PR.

- I think it would make sense if the optimizer (e.g. LBFGS in this case) accumulated the cost function, gradient etc. call counts into its own call counts because these are the numbers that are reported to an observer. Currently the reported numbers are misleading because the functions are actually called much more often depending on the line search.
- Currently the backtracking line search evaluates its condition (in [`terminate`](https://github.com/argmin-rs/argmin/blob/f5ffe75979f5d14574e666837bd1220c2e28e727/src/solver/linesearch/backtracking.rs#L163)) at iteration zero (i.e. directly after the call to `init` and before the first call to `next_iter`) without actually applying the search direction to the current state. This doesn't make any sense to me.
- The `alpha` values in backtracking line search are actually "shifted" by one application of `rho` when passed into the condition. For example in the first iteration (`iter == 1`) it actually checks the condition with a cost and gradient evaluated at `x + alpha * search_direction` (where `alpha == initial_alpha`) but at the same time supplies the value [`alpha = initial_alpha * rho`](https://github.com/argmin-rs/argmin/blob/f5ffe75979f5d14574e666837bd1220c2e28e727/src/solver/linesearch/backtracking.rs#L150) to the condition. This appears to be inconsistent.

I'm not sure if these issues apply to other line searches as well. If you agree that these points should be addressed I can submit a PR.

**Comment by stefan-k** [comment]
> * I think it would make sense if the optimizer (e.g. LBFGS in this case) accumulated the cost function, gradient etc. call counts into its own call counts because these are the numbers that are reported to an observer. Currently the reported numbers are misleading because the functions are actually called much more often depending on the line search.

This is actually what `op.consume_op(line_op);` in [line 162 of LBFGS is supposed to do](https://github.com/argmin-rs/argmin/blob/f5ffe75979f5d14574e666837bd1220c2e28e727/src/solver/quasinewton/lbfgs.rs#L162). I'm surprised that this doesn't work. 

> * Currently the backtracking line search evaluates its condition (in [`terminate`](https://github.com/argmin-rs/argmin/blob/f5ffe75979f5d14574e666837bd1220c2e28e727/src/solver/linesearch/backtracking.rs#L163)) at iteration zero (i.e. directly after the call to `init` and before the first call to `next_iter`) without actually applying the search direction to the current state. This doesn't make any sense to me.

At first I thought this should have no effect because the termination condition should not be met at this point. But if I understand this correctly the problem is that `alpha` at that point is not consistent with the step that has not been taken. Is this correct?

> * The `alpha` values in backtracking line search are actually "shifted" by one application of `rho` when passed into the condition. For example in the first iteration (`iter == 1`) it actually checks the condition with a cost and gradient evaluated at `x + alpha * search_direction` (where `alpha == initial_alpha`) but at the same time supplies the value [`alpha = initial_alpha * rho`](https://github.com/argmin-rs/argmin/blob/f5ffe75979f5d14574e666837bd1220c2e28e727/src/solver/linesearch/backtracking.rs#L150) to the condition. This appears to be inconsistent.

This is in fact inconsistent. A potential solution to both issues could be to perform the initial full step in the `init(...)` method and then move the update of alpha to the beginning of `next_iter(...)`. 

I would appreciate a PR.

**Comment by w1th0utnam3** [comment]
> This is actually what `op.consume_op(line_op);` in [line 162 of LBFGS is supposed to do](https://github.com/argmin-rs/argmin/blob/f5ffe75979f5d14574e666837bd1220c2e28e727/src/solver/quasinewton/lbfgs.rs#L162). I'm surprised that this doesn't work.
Ah I see, sorry I forgot to check what this call actually does. Apparently the counts in the operator returned by the `Executor` are all zero. The returned `IterState` however contains the information. I'm not sure how this `OpWrapper` concept is working so I'm a bit confused right now.

> At first I thought this should have no effect because the termination condition should not be met at this point. But if I understand this correctly the problem is that `alpha` at that point is not consistent with the step that has not been taken. Is this correct?

Right, this is probably the problem in my case.

> This is in fact inconsistent. A potential solution to both issues could be to perform the initial full step in the `init(...)` method and then move the update of alpha to the beginning of `next_iter(...)`.

Right, this makes sense. Although I think this way isn't really possible because `init(...)` only receives a non-mutable reference to the state? For now I worked around this by adding `state.iter > 0` to the termination check and a field `next_alpha` to the solver which is initialized to `alpha` and in the `next_iter(...)` function you first take this value `alpha = next_alpha` and then update it `next_alpha = alpha * rho`.

**Comment by w1th0utnam3** [comment]
Ah sorry, `init` can return an optional `ArgminIterData`, this can be used of course. Sorry, overlooked that.

**Comment by stefan-k** [comment]
> I'm not sure how this `OpWrapper` concept is working so I'm a bit confused right now.

Basically it only acts as a layer in between the solver and the struct implementing `ArgminOp` in order to count the function evaluations. Via `consume_op` it adds the number of function evaluations of two `OpWrapper`s.

> Apparently the counts in the operator returned by the `Executor` are all zero. The returned `IterState` however contains the information. 

It is strange that you see a discrepancy between the two. If the linesearch ends prematurely due to the "termination bug" that you found, then you would expect all function counts within the linesearch to be zero (because cost function value and gradient are supplied by LBFGS and not computed again in the backtracking linesearch). 

> Ah sorry, init can return an optional ArgminIterData, this can be used of course. Sorry, overlooked that.

Yes I think the optional return value of `init(...)` is the way to go.

**Comment by w1th0utnam3** [comment]
Ok, I opened a PR for the `alpha` issue.

> It is strange that you see a discrepancy between the two. If the linesearch ends prematurely due to the "termination bug" that you found, then you would expect all function counts within the linesearch to be zero (because cost function value and gradient are supplied by LBFGS and not computed again in the backtracking linesearch).

This also happens when the linesearch actually performs multiple iterations and finds a solution. Inside of the backtracking linesearch I can print the current value of `op.cost_func_count` and can see it increasing from 0 to the number of iterations required by the linesearch. But then the `cost_func_count` of the `OpWrapper` returned by the `Executor` only contains zeros and therefore the `consume` does nothing.

**Comment by w1th0utnam3** [comment]
So in LBFGS it looks like this:
```rust
let ArgminResult {
            operator: line_op,
            state:
                IterState {
                    param: xk1,
                    cost: next_cost,
                    ..
                },
        } = Executor::new(
            OpWrapper::new_from_wrapper(op),
            self.linesearch.clone(),
            param.clone(),
        )
        .grad(prev_grad.clone())
        .cost(cur_cost)
        .ctrlc(false)
        .run()?;
```
and in the end of `Executor::run` we have
```rust
Ok(ArgminResult::new(self.op.get_op(), self.state))
```

The `OpWrapper::new_from_wrapper(op)` basically adds a layer so the we can count the evaluations of the line search on its own. But in the end of `Executor::run` this information is discarded by the `self.op.get_op()` call as it only returns the `op` not the `OpWrapper` which actually contains the evaluation counts of the linesearch. As I'm not familiar with the architecture of the rest of the library, I don't know what the correct solution would be here.

**Comment by stefan-k** [comment]
> Ok, I opened a PR for the `alpha` issue.

Thanks a lot, I will have a look soon!

> This also happens when the linesearch actually performs multiple iterations and finds a solution. Inside of the backtracking linesearch I can print the current value of `op.cost_func_count` and can see it increasing from 0 to the number of iterations required by the linesearch. But then the `cost_func_count` of the `OpWrapper` returned by the `Executor` only contains zeros and therefore the `consume` does nothing.

This is really strange. I had a quick look but couldn't find anything obvious. It might be rooted [here](https://github.com/argmin-rs/argmin/blob/f5ffe75979f5d14574e666837bd1220c2e28e727/src/core/executor.rs#L229) in the line `Ok(ArgminResult::new(self.op.get_op(), self.state))`. The operator is moved out via `self.op.get_op()`, effectively removing one `OpWrapper` layer. But I believe there are two layers and the underlying `OpWrapper` layer should also have the function counts (Having two layers isn't really the best design admittedly).
You're welcome to dig into this if you want, otherwise I will try to have a look in the next couple of days. This certainly needs to be fixed.

**Comment by stefan-k** [comment]
Sorry I missed your post when submitting mine and therefore posted something quite similar to you ;) 

> The `OpWrapper::new_from_wrapper(op)` basically adds a layer so the we can count the evaluations of the line search on its own. But in the end of `Executor::run` this information is discarded by the `self.op.get_op()` call as it only returns the `op` not the `OpWrapper` which actually contains the evaluation counts of the linesearch. As I'm not familiar with the architecture of the rest of the library, I don't know what the correct solution would be here.

Since `consume_op` only takes an `OpWrapper<O>`, I'm assuming that it must be two layers of `OpWrapper`, and then the function counts should also be at the bottom-most layer. But I'm fairly certain that the problem must be somewhere in that area.

**Comment by stefan-k** [comment]
I provided a fix in #156. I'm still not completely happy with the `OpWrapper` design, but I think the mentioned problem should be fixed. I would appreciate it if you could give it a try!

**Comment by w1th0utnam3** [comment]
Yes, that seems to work. Thanks!
